staging/src/github.com/openshift/oc/pkg/cli/version/version.go

```diff
@@ -1,25 +1,199 @@
+	)
+	if !o.ServerOnly {
+		clientVersion := version.Get()
+		versionInfo.ClientVersion = &clientVersion
+	}
+	if !ko.ClientOnly {
+		if o.discoveryClient != nil {
+			// Always request fresh data from the server
+			o.discoveryClient.Invalidate()
+			serverVersion, serverErr = o.discoveryClient.ServerVersion()
+			versionInfo.ServerVersion = serverVersion
+		}
+		if o.oClient != nil {
+			clusterOperator, serverErr = o.oClient.ClusterOperators().Get("openshift-apiserver", metav1.GetOptions{})
+			// if error here, log and move on
+			if serverErr != nil {
+				switch {
+				case kerrors.IsForbidden(serverErr):
+					// this occurs when a non-admin user runs `oc version` since only admins
+					// can gather info from clusteroperators
+					fmt.Fprintf(o.Out, "OpenShift Server Version: Unauthorized\n")
+					serverErr = nil
+				case kerrors.IsNotFound(serverErr):
+					// this occurs when server can not find
+					// clusteroperators.config.openshift.io openshift-apiserver
+					fmt.Fprintf(o.Out, "OpenShift Server Version: Not Found\n")
+					serverErr = nil
+				}
+			}
+			if clusterOperator != nil {
+				for _, ver := range clusterOperator.Status.Versions {
+					if ver.Name == "operator" {
+						// openshift-apiserver does not report version,
+						// clusteroperator/openshift-apiserver does, and only version number
+						versionInfo.OpenShiftVersion = ver.Version
+					}
+				}
+			}
+		}
+	}
+
+	switch ko.Output {
+	case "":
+		if ko.Short {
+			if len(versionInfo.OpenShiftVersion) != 0 {
+				fmt.Fprintf(o.Out, "OpenShift Server Version: %s\n", fmt.Sprintf("%s", versionInfo.OpenShiftVersion))
+			}
+			if versionInfo.ClientVersion != nil {
+				fmt.Fprintf(o.Out, "OpenShift Client Version: %s\n", versionInfo.ClientVersion.GitVersion)
+			}
+			if versionInfo.ServerVersion != nil {
+				fmt.Fprintf(o.Out, "Kubernetes Server Version: %s\n", versionInfo.ServerVersion.GitVersion)
+			}
+		} else {
+			if len(versionInfo.OpenShiftVersion) != 0 {
+				fmt.Fprintf(o.Out, "OpenShift Server Version: %s\n", fmt.Sprintf("%s", versionInfo.OpenShiftVersion))
+			}
+			if versionInfo.ClientVersion != nil {
+				fmt.Fprintf(o.Out, "OpenShift Client Version: %s\n", fmt.Sprintf("%#v", versionInfo.ClientVersion))
+			}
+			if versionInfo.ServerVersion != nil {
+				fmt.Fprintf(o.Out, "Kubernetes Server Version: %s\n", fmt.Sprintf("%#v", versionInfo.ServerVersion))
+			}
+		}
+	case "yaml":
+		marshalled, err := yaml.Marshal(&versionInfo)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(o.Out, string(marshalled))
+	case "json":
+		marshalled, err := json.MarshalIndent(&versionInfo, "", "  ")
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(o.Out, string(marshalled))
+	default:
+		// There is a bug in the program if we hit this case.
+		// However, we follow a policy of never panicking.
+		return fmt.Errorf("VersionOptions were not validated: --output=%q should have been rejected", ko.Output)
+	}
+
+	return serverErr
```

Comment:
this looks ok'ish. The alternative to reuse the upstream command object and extend the output does not work? Just asking. If it is too complex and ugly, we have to go the way of copying.

Reply:
I think we need to add the newCmd to mimic/copy upstream at this point.  We need to change the output not simply extend it (we'll want 'Kubernetes Server Version', 'OpenShift Client Version'), add the logic for OpenShift apiserver version.  I'll spend a bit more time looking at it, though.

Reply:
On the other hand, do we really want to keep the cmd output exactly the same as in 3.x, but simply add the Openshift api version?  
If so, we'd have
```
Client Version: blah
Server Version: blah
OpenShift Server Version: blah
``` 
That will keep things closest to 3.x while adding the required info, but not as descriptive/clear.

Reply:
Final verdict for now is to keep with 3.x output, and add OpenShift Version, if and only if one is logged in as admin to a cluster.  Output identical to 3.x if OpenShift Version not found (non-admins)